sv.c

```diff
@@ -6788,8 +6789,10 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
                  * set SvLEN to whatever value was in the now-freed
                  * regex body. The PVX buffer is shared by multiple re's
                  * and only freed once, by the re whose len in non-null */
-                STRLEN len = ReANY(sv)->xpv_len;
+                struct regexp *r = ((XPV*)SvANY(sv))->xpv_len_u.xpvlenu_rx;
```

Comment:
Unless I've missed a reason, I think this should stick with `struct regexp *r = ReANY(sv);` - I see no benefit to manually expanding the inline function, and some risk if `ReANY` changes again in the future.

Also in the commit message,  I think you intended to follow `with commit df6b4bd56551f2d3:` with that commit's subject line `give REGEXP SVs the POK flag again`.

Reply:
> Unless I've missed a reason, I think this should stick with `struct regexp *r = ReANY(sv);` - I see no benefit to manually expanding the inline function, and some risk if `ReANY` changes again in the future.

The diff hunk doesn't really give enough context. What happened was..

I started with code in `sv.c` that reads like this (in blead)

```
            if (isREGEXP(sv)) {
                /* SvLEN points to a regex body. Free the body, then
                 * set SvLEN to whatever value was in the now-freed
                 * regex body. The PVX buffer is shared by multiple re's
                 * and only freed once, by the re whose len in non-null */
                STRLEN len = ReANY(sv)->xpv_len;
                pregfree2((REGEXP*) sv);
                SvLEN_set((sv), len);
                goto freescalar;
            }
```

and I read it, and thought "wait, what's going on here. The comment mentions `SvLEN` but I don't see any sign of `SvLEN`."

So then I had to figure out that the `ReANY` is actually what the `SvLEN` is talking about in the comment - it's not obvious

And then I had to go look up the definition of that

```
PERL_STATIC_INLINE struct regexp *
Perl_ReANY(const REGEXP * const re)
{
    XPV* const p = (XPV*)SvANY(re);

    PERL_ARGS_ASSERT_REANY;
    assert(isREGEXP(re));

    return SvTYPE(re) == SVt_PVLV ? p->xpv_len_u.xpvlenu_rx
                                   : (struct regexp *)p;
}
```

at which point I thought "eh, what's that ternary doing? Sometimes it isn't `SvLEN`"

So then I had to re-read the code in `sv.c`.

Enough to look for more context:

```
        case SVt_PVLV:
            if (LvTYPE(sv) == 'T') { /* for tie: return HE to pool */
                SvREFCNT_dec(HeKEY_sv((HE*)LvTARG(sv)));
                HeNEXT((HE*)LvTARG(sv)) = PL_hv_fetch_ent_mh;
                PL_hv_fetch_ent_mh = (HE*)LvTARG(sv);
            }
            else if (LvTYPE(sv) != 't') /* unless tie: unrefcnted fake SV**  */
                SvREFCNT_dec(LvTARG(sv));
            if (isREGEXP(sv)) {
                /* SvLEN points to a regex body. Free the body, then
                 * set SvLEN to whatever value was in the now-freed
                 * regex body. The PVX buffer is shared by multiple re's
                 * and only freed once, by the re whose len in non-null */
                STRLEN len = ReANY(sv)->xpv_len;
                pregfree2((REGEXP*) sv);
                SvLEN_set((sv), len);
                goto freescalar;
            }
```

and then I had enough to mentally simplify the code in `Perl_ReANY` - it was correct, and it always took one side of the ternary.

And that was a bunch of work, involving two context switches. I didn't find this very readable.

Hence why I deliberately decided to inlinine - based on this experience I felt that it made the code more readable:

```
            if (isREGEXP(sv)) {
                /* SvLEN points to a regex body. Free the body, then
                 * set SvLEN to whatever value was in the now-freed
                 * regex body. The PVX buffer is shared by multiple re's
                 * and only freed once, by the re whose len in non-null */
                struct regexp *r = ((XPV*)SvANY(sv))->xpv_len_u.xpvlenu_rx;
                STRLEN len = r->xpv_len;
                pregfree2((REGEXP*) sv);
                del_body_by_type(r, SVt_REGEXP);
                SvLEN_set((sv), len);
                goto freescalar;
            }
 ```


I also don't think that there's real risk here - this code is "inside" the abstraction of how regexs are store in two possible places, given that it's responsible for correctly freeing up one of those two possible places. If `ReANY` needs to change, then this code needs to be changed too, or at least audited that it's still correct.


So that's how I saw it. It didn't feel like using `ReANY` worked as an abstraction, because it didn't seem to be something you could ignore while trying to understand (or verify) the code - it was getting in the way, rather than making it simpler.

Reply:
> If `ReANY` needs to change, then this code needs to be changed too, or at least audited that it's still correct.

In such circumstance, I'd have imagined one would primarily rely on grepping for ReANY to find what needs auditing. I'd suggest at least mentioning it in the comment: `SvLEN points to a regex body (the PVLV case in ReANY).`.

(More formally, I think it dubious to assert that SvLEN does the pointing, but I guess that's a reasonable enough shorthand.)

Reply:
> Quite a lot of comments. Do these work better than what we had?

I think so, yes.

> If you think that they are good enough to merge, as part of the rebase I was proposing to re-order the commits so that this commit with comments was immediately after the 2 commits with code changes, and before the commits that only deal with `.github/workflows/testsuite.yml`

That makes sense to me.

Reply:
We don't seem to be getting more eyeballs.

Even if we punt on the workflow changes until we get a better answer on "quota", I'd quite like to get the 3 code commits into blead, because they make the testsuite 100% clean under ASAN (ie no leaks). Right now every time I've done an ASAN build on blead or anything branched from it, I've had to re-confirm that the 3 tests I see failing are the 3 that I know "should" fail. Getting blead clean (and keeping it that way) would be good.

Reply:
> We don't seem to be getting more eyeballs.
> 
> Even if we punt on the workflow changes until we get a better answer on "quota", I'd quite like to get the 3 code commits into blead

I'd go for it, we're early enough in the release cycle to have lots of time.